packages/plugin-http-client/src/rest/BaseEntity.ts

```diff
@@ -0,0 +1,196 @@
+import { clone } from '@ima/helpers';
+
+import { BaseMapper, DataFieldValue, MapperItem } from './mapper/BaseMapper';
+
+export interface EntityConstructor {
+  new (data: object): Entity;
+}
+
+export interface Entity {
+  serialize(data: object): any;
+  deserialize(data: object): any;
+}
+
+/**
+ * The base class for typed REST API entities. Usage of typed entities may be
+ * optional and is dependent on the specific implementation of the REST API
+ * client.
+ */
+export class BaseEntity {
+  /**
+   * Returns the description of automatic mapping of the raw data exchanged
+   * between the REST API, and the properties of this
+   * entity.
+   *
+   * The keys of the returned object are the names of the API properties.
+   *
+   * The value associated with key is one of the following:
+   *
+   * - The new name of the property in this entity. This is useful when the
+   *   property only needs to be renamed.
+   *
+   * - The mapper instance of BaseMapper. This is useful when the property only needs to changed format but not renamed.
+   *
+   * - The MapperItem object eg.: { mapper: new EntityListMapper(BaseEntity), newKey: 'authors' }.
+   *  This is useful when the property needs changed format and renamed too.
+   *
+   * @returns {Object<string, (
+   *          string |
+   *          Mappper |
+   *           {
+   *             newKey: string,
+   *             mapper: BaseMapper,
+   *           }
+   *         )>} The description of how the raw data properties should be
+   *         mapped to the entity properties and vice versa.
+   */
+  get dataFieldMapping(): {
+    [key: string]: DataFieldValue;
+  } {
+    return {};
+  }
+
+  /**
+   * Initializes the entity.
+   *
+   * @param {Object<string, *>} data Entity data, which will be directly
+   *        assigned to the entity's fields.
+   */
+  constructor(data: object) {
+    const entityData = this.deserialize(data);
+
+    Object.assign(this, entityData);
+  }
+
+  /**
+   * Serializes this entity into a JSON-compatible plain JavaScript object
+   * that has a structure that is compatible with the entity's REST API
+   * resource.
+   *
+   * Note that this method may not receive a representation of the entity's
+   * complete state if invoked by the {@linkcode cloneAndPatch()} method.
+   *
+   * The default implementation of this method implements a mapping based on
+   * the {@linkcode dataFieldMapping} property's value.
+   *
+   * @param {BaseEntity} data
+   * @returns {Object<string, *>} Data object representing this entity in a
```

Comment:
Nevím jak moc jsou tyhle JSdocy valid v rámci typescriptu... klidně bych je i smazal, protože v JSdocu to typově omezuješ na něco, ale v parametru je potom něco jiného.

Tzn vlastně v podstatě člověk potom neví co si má vybrat :D osobně bych to smazal, kdyby se někdy použil TSdoc na generování tak to stejně vyhodí jako chybu.

Reply:
Odmazano.